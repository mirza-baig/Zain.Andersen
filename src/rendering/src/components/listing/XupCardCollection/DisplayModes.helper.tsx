import { ComponentRendering, Placeholder } from '@sitecore-jss/sitecore-jss-nextjs';
import classNames from 'classnames';
import React, { useRef } from 'react';
import { SliderWrapper } from 'src/helpers/SliderWrapper';
import { getLayoutClasses } from './XupCardCollection.helper';
import { DesktopVideoDisplayStyleType } from './XupCardCollection';
import { SliderRefType, sliderSettings, SliderType } from 'src/helpers/SliderWrapper/SliderWrapper';

type FavoriteProducts = {
  favoriteProducts: string[];
};
type RenderingProps = {
  rendering: ComponentRendering;
};

type GridDisplayProps = {
  maxCardsPerRow?: number;
  desktopVideoDisplayStyle: DesktopVideoDisplayStyleType;
} & RenderingProps &
  FavoriteProducts;

type SliderDisplayProps = {
  desktopVideoDisplayStyle: DesktopVideoDisplayStyleType;
  sliderSettings: sliderSettings;
} & RenderingProps &
  FavoriteProducts;

export const GridDisplay = ({
  rendering,
  maxCardsPerRow,
  favoriteProducts,
  desktopVideoDisplayStyle,
}: GridDisplayProps) => {
  return (
    <Placeholder
      name="cards"
      rendering={rendering}
      favoriteProducts={favoriteProducts}
      desktopVideoDisplayStyle={desktopVideoDisplayStyle}
      render={(cards) =>
        cards.map((card, index) => (
          // console.log('card', card),
          <div
            key={index}
            className={classNames(
              'col-span-12',
              maxCardsPerRow && getLayoutClasses(maxCardsPerRow)
            )}
          >
            {card}
          </div>
        ))
      }
    />
  );
};

export const SliderDisplay = ({
  rendering,
  sliderSettings,
  favoriteProducts,
  desktopVideoDisplayStyle,
}: SliderDisplayProps) => {
  const sliderRef = useRef<SliderType>(null);

  return (
    <div className="col-span-12 [&_.slick-track]:!mx-0">
      <Placeholder
        name="cards"
        rendering={rendering}
        favoriteProducts={favoriteProducts}
        desktopVideoDisplayStyle={desktopVideoDisplayStyle}
        render={(cards) => (
          <SliderWrapper sliderSettings={sliderSettings} sliderRef={sliderRef as SliderRefType}>
            {cards?.map((card, index) => (
              <div
                onKeyDown={(e) => {
                  if (e.shiftKey && e.keyCode == 9) {
                    sliderRef?.current && sliderRef.current.slickPrev();
                  } else if (e.keyCode == 9) {
                    sliderRef?.current && sliderRef?.current.slickNext();
                  }
                }}
                tabIndex={0}
                key={index}
                className={classNames('h-full md:px-xxs')}
              >
                {card}
              </div>
            ))}
          </SliderWrapper>
        )}
      />
    </div>
  );
};
