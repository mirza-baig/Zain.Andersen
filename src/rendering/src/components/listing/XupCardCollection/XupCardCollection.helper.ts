export const ColumnSpans: { [cardsInRow: number]: string } = {
  1: 'md:col-span-12',
  2: 'md:col-span-6',
  3: 'md:col-span-4',
  4: 'md:col-span-3',
};

export const getLayoutClasses = (maxCardsPerRow: number): string => {
  return ColumnSpans[maxCardsPerRow];
};
