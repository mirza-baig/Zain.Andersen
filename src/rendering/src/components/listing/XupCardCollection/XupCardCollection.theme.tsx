import { ThemeFile } from 'lib/context/ThemeContext';

export const XupCardCollectionTheme: ThemeFile = {
  aw: {
    classes: {
      headlineClass: 'text-sm-m md:text-m font-bold mb-xxs',
      bodyClass: 'text-body text-dark-gray font-regular mb-m',
      buttonGroupClass: {
        wrapper: 'flex flex-col md:flex-row items-start',
      },
    },
  },
  rba: {
    classes: {
      headlineClass: 'text-m md:text-l font-medium mb-xxs',
      bodyClass: 'text-body text-dark-gray font-regular mb-m',
      buttonGroupClass: {
        wrapper: 'flex flex-col md:flex-row items-start',
      },
    },
  },
};
