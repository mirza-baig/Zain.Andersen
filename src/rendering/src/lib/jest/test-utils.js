// Global
import { render, fireEvent, waitFor } from '@testing-library/react';
import { ComponentPropsContext } from '@sitecore-jss/sitecore-jss-nextjs';
import { FastSitecoreContext } from 'lib/overrides/FastSitecoreContext';

// Lib
import { createSampleComponentFactory } from 'lib/mocks/mock-placeholder';

export const eeSitecoreContext = {
  route: {
    name: 'Foo',
    placeholders: {},
  },
  context: {
    pageEditing: true,
    site: {
      name: 'JssTestWeb',
    },
    language: 'en',
  },
};

export const renderComponent = (
  Component,
  {
    componentProps = {},
    staticProps = {},
    componentFactory = createSampleComponentFactory(),
    sitecoreContext = {
      route: {
        name: 'Foo',
        placeholders: {},
      },
    },
  }
) => {
  const mockLayoutData = {
    sitecore: sitecoreContext,
  };

  return render(
    <ComponentPropsContext value={staticProps}>
      <FastSitecoreContext componentFactory={componentFactory} layoutData={mockLayoutData}>
        <Component {...componentProps} />
      </FastSitecoreContext>
    </ComponentPropsContext>
  );
};

export const snapshot = (Component, mockData) => {
  const component = renderComponent(Component, mockData);
  expect(component.container.firstChild).toMatchSnapshot();
  return component;
};

export const hasDataComponent = async (component, name) => {
  await waitFor(() => {
    const nodes = component.baseElement.querySelectorAll(`[data-component="${name}"]`);
    return expect(nodes.length).toBe(1);
  });
};

export const mockDateAs = (epochMs) => {
  if (typeof epochMs !== 'number') {
    console.error(
      'mockDateAs only accepts epoch values in milliseconds to prevent timezones from factoring in to Date creation.'
    );
  }

  jest.spyOn(global.Date, 'now').mockImplementation(() => new Date(epochMs).valueOf());
};

export const testFieldError = async (Component, mockData, { label, value, error }) => {
  const component = renderComponent(Component, mockData);
  const field = component.getByLabelText(label);
  fireEvent.change(field, { target: { value } });
  field.blur();
  await waitFor(() => {
    const err = component.queryByText(error);
    return expect(err).not.toBe(null);
  });
};
