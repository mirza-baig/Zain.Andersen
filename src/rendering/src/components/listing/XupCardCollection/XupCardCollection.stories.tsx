// Global
import { Story, Meta } from '@storybook/react';
// Local
import XupCardCollection, { XupCardCollectionProps } from './XupCardCollection';
import defaultData from './XupCardCollection.mock-data';

export default {
  title: 'Listing/XupCardCollection',
  component: XupCardCollection,
} as Meta;
const Template: Story<XupCardCollectionProps> = (props) => <XupCardCollection {...props} />;
export const Default = Template.bind({});
Default.args = defaultData;
