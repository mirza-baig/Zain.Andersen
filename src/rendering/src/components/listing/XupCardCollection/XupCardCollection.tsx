// Components
import { withDatasourceCheck } from '@sitecore-jss/sitecore-jss-nextjs';
import classNames from 'classnames';
import { useTheme } from 'lib/context/ThemeContext';
import { getEnum } from 'lib/utils';
import { getBreakpoint, useCurrentScreenType } from 'lib/utils/get-screen-type';
import { ReactElement } from 'react';
import { Feature } from 'src/.generated/Feature.EnterpriseWeb.model';
import { Component } from 'src/helpers/Component';
import { GridDisplay, SliderDisplay } from './DisplayModes.helper';
import { XupCardCollectionTheme } from './XupCardCollection.theme';
import { BodyCopy } from 'src/helpers/BodyCopy';
import { ButtonGroup } from 'src/helpers/ButtonGroup';
import { Headline } from 'src/helpers/Headline';
import { DefaultEditFrameButton, EditFrame } from 'src/helpers/EditFrame';

export type XupCardCollectionProps =
  Feature.EnterpriseWeb.Enterprise.Components.Listing.XupCardCollection.XupCardContainer;

export type XupDisplayStyle = 'grid' | 'horizontal-scroll';

export type DesktopVideoDisplayStyleType = 'in-line' | 'in-modal';

const XupCardCollection = (props: XupCardCollectionProps): JSX.Element => {
  const { currentScreenWidth } = useCurrentScreenType();
  const { themeData } = useTheme(XupCardCollectionTheme);

  const desktopDisplayStyle =
    getEnum<XupDisplayStyle>(props?.fields?.desktopDisplayStyle) || 'grid';
  const mobileDisplayStyle = getEnum<XupDisplayStyle>(props?.fields?.mobileDisplayStyle) || 'grid';

  const maxCardsPerRow: number = parseInt(getEnum<string>(props?.fields?.cardsPerRow) || '3');
  const desktopVideoDisplayStyle =
    getEnum<DesktopVideoDisplayStyleType>(props.fields?.desktopVideoDisplayStyle) || 'in-line';

  const sliderSettings = {
    dots: false,
    infinite: false,
    slidesToShow: maxCardsPerRow,
    responsive: [
      {
        breakpoint: getBreakpoint('mmd'),
        settings: {
          slidesToShow: maxCardsPerRow >= 4 ? 3 : maxCardsPerRow,
          dots: true,
        },
      },
      {
        breakpoint: getBreakpoint('md'),
        settings: {
          slidesToShow: 1,
          dots: true,
        },
      },
    ],
  };

  const renderCardColection = (): ReactElement => {
    if (currentScreenWidth < getBreakpoint('md')) {
      // Renderings for mobile devices
      if (mobileDisplayStyle === 'grid') {
        return (
          <GridDisplay
            rendering={props?.rendering}
            desktopVideoDisplayStyle={desktopVideoDisplayStyle}
            favoriteProducts={props?.favoriteProducts}
          />
        );
      } else {
        return (
          <SliderDisplay
            rendering={props?.rendering}
            sliderSettings={sliderSettings}
            favoriteProducts={props?.favoriteProducts}
            desktopVideoDisplayStyle={desktopVideoDisplayStyle}
          />
        );
      }
    } else {
      // Renderings for tablets and large screen devices
      if (desktopDisplayStyle === 'grid') {
        return (
          <GridDisplay
            rendering={props?.rendering}
            maxCardsPerRow={maxCardsPerRow}
            favoriteProducts={props?.favoriteProducts}
            desktopVideoDisplayStyle={desktopVideoDisplayStyle}
          />
        );
      } else {
        return (
          <SliderDisplay
            rendering={props?.rendering}
            sliderSettings={sliderSettings}
            favoriteProducts={props?.favoriteProducts}
            desktopVideoDisplayStyle={desktopVideoDisplayStyle}
          />
        );
      }
    }
  };

  return (
    <EditFrame
      title="Xup Card Collection Cards"
      dataSource={{ itemId: props?.rendering?.dataSource }}
      buttons={[{ ...DefaultEditFrameButton.edit, fields: ['ph-cards'] }]}
    >
      <Component variant="lg" dataComponent="listing/xupcardcollection" {...props}>
        <div className={classNames('col-span-12')}>
          <Headline classes={themeData.classes.headlineClass} {...props} />
          <BodyCopy classes={themeData.classes.bodyClass} {...props} />
          <ButtonGroup classes={themeData.classes.buttonGroupClass} {...props} />
        </div>
        {renderCardColection()}
      </Component>
    </EditFrame>
  );
};

export default withDatasourceCheck()<XupCardCollectionProps>(XupCardCollection);
