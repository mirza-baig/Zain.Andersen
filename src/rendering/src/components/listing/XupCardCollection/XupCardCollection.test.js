// Global
import { hasDataComponent, snapshot } from 'lib/jest/test-utils';
import 'lib/jest/match-media-mock';
// Local
import XupCardCollection from './XupCardCollection';
import defaultData from './XupCardCollection.mock-data';

it('renders correctly', async () => {
  const component = snapshot(XupCardCollection, { componentProps: defaultData });
  await hasDataComponent(component, 'listing/xupcardcollection');
});
