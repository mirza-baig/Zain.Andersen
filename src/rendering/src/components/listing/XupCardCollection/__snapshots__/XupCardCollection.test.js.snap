// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<section
  data-component="listing/xupcardcollection"
  id="XupCardCollection"
>
  <div
    class="my-8 lg:mx-auto md:max-w-screen-lg theme-white"
  >
    <div
      class="section-grid grid grid-cols-2 md:grid-cols-12 px-m gap-s md:gap-s "
    >
      <div
        class="col-span-12"
      >
        <div
          class="items-top font-sans [&_a:hover]:underline text-sm-m md:text-m font-bold mb-xxs"
        >
          <h2>
            Xup Card Collection
          </h2>
        </div>
        <div
          class="text-body text-dark-gray font-regular mb-m"
        >
          <div
            class="body-copy"
          >
            <div>
              Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nunc vulputate libero et velit interdum, ac aliquet odio mattis. Class aptent taciti sociosqu ad litora torquent per conubia nostra, per inceptos himenaeos. Curabitur tempus urna at turpis condimentum lobor.
            </div>
          </div>
        </div>
        <div
          class="flex flex-col md:flex-row items-start mb-s flex items-start md:flex-row md:items-center md:space-x-4"
        >
          <a
            aria-label="Call to Action"
            class="hover:decoration-theme-btn-decoration flex w-fit items-center whitespace-normal font-sans text-text-link font-heavy text-theme-text hover:underline hover:underline-offset-8 disabled:border-gray disabled:text-gray mr-m"
            data-gtm-click=""
            data-gtm-dl-event="cta_click"
            href="/"
            role="link"
            tabindex="0"
            title=""
          >
            Call to Action
            <span
              class="ml-xxs"
            />
          </a>
        </div>
      </div>
      <div
        class="col-span-12 md:col-span-4"
      >
        <div
          style="background: darkorange; outline: 5px solid orange; padding: 10px; color: white; max-width: 500px;"
        >
          <h2>
            GenericCard
          </h2>
          <p>
            JSS component is missing React implementation. See the developer console for more information.
          </p>
        </div>
      </div>
      <div
        class="col-span-12 md:col-span-4"
      >
        <div
          style="background: darkorange; outline: 5px solid orange; padding: 10px; color: white; max-width: 500px;"
        >
          <h2>
            GenericCard
          </h2>
          <p>
            JSS component is missing React implementation. See the developer console for more information.
          </p>
        </div>
      </div>
      <div
        class="col-span-12 md:col-span-4"
      >
        <div
          style="background: darkorange; outline: 5px solid orange; padding: 10px; color: white; max-width: 500px;"
        >
          <h2>
            GenericCard
          </h2>
          <p>
            JSS component is missing React implementation. See the developer console for more information.
          </p>
        </div>
      </div>
      <div
        class="col-span-12 md:col-span-4"
      >
        <div
          style="background: darkorange; outline: 5px solid orange; padding: 10px; color: white; max-width: 500px;"
        >
          <h2>
            GenericCard
          </h2>
          <p>
            JSS component is missing React implementation. See the developer console for more information.
          </p>
        </div>
      </div>
      <div
        class="col-span-12 md:col-span-4"
      >
        <div
          style="background: darkorange; outline: 5px solid orange; padding: 10px; color: white; max-width: 500px;"
        >
          <h2>
            GenericCard
          </h2>
          <p>
            JSS component is missing React implementation. See the developer console for more information.
          </p>
        </div>
      </div>
      <div
        class="col-span-12 md:col-span-4"
      >
        <div
          style="background: darkorange; outline: 5px solid orange; padding: 10px; color: white; max-width: 500px;"
        >
          <h2>
            GenericCard
          </h2>
          <p>
            JSS component is missing React implementation. See the developer console for more information.
          </p>
        </div>
      </div>
      <div
        class="col-span-12 md:col-span-4"
      >
        <div
          style="background: darkorange; outline: 5px solid orange; padding: 10px; color: white; max-width: 500px;"
        >
          <h2>
            GenericCard
          </h2>
          <p>
            JSS component is missing React implementation. See the developer console for more information.
          </p>
        </div>
      </div>
      <div
        class="col-span-12 md:col-span-4"
      >
        <div
          style="background: darkorange; outline: 5px solid orange; padding: 10px; color: white; max-width: 500px;"
        >
          <h2>
            GenericCard
          </h2>
          <p>
            JSS component is missing React implementation. See the developer console for more information.
          </p>
        </div>
      </div>
    </div>
  </div>
</section>
`;
